### PR TITLE
improved likelyhood of dropping into catch when checking for thenables

### LINKED ISSUE #2: Consider dropping the try-catch
### Background

_ruply_ is promise-aware. Instead of passing promises into your callbacks, it awaits them and passes the values to which they resolve instead. This is what the developer intends in almost every scenario.
```javascript
return run(
	database.query(`
		select
			id, title
		from
			books
	`),
	rows => new Map(
		rows.map(({ id, title }) => [id, title])
	)
)
```

In the world of JavaScript ‒ however ‒ "is this a promise?" can be a trickier question than it initially seems.

[ECMAScript has `IsPromise`](https://tc39.es/ecma262/#sec-ispromise) for this purpose. It checks for the existence of an internal slot. _ruply_ can't do that.

It could instead do `candidate instanceof Promise`. This has the downside of discriminating against "fake" promises. Non-globally polyfilled promises would not be considered promises (you may not be using a polyfill, but a library or framework you're using might!). "Fake" promises designed for testing, or logging, or any other purpose, would not be considered promises either. Finally, I feel this goes against the duck typed nature of JavaScript.

[Promises/A+ describes _thenables_](https://promisesaplus.com/#terminology), defined as:
 > an object or function that defines a `then` method.

This definition covers ECMA-style promises, as well as anything else one might expect to be handled like a promise.

_ruply_ currently has `checkThenable`:
```javascript
function checkThenable(input) {
	try {
		var { then } = input
	} catch (error) {
		return false
	}
	return 'function' == typeof then
}
```
The try-catch construction ensures `null` and `undefined` are accurately reported as not being promises, as well as objects like this one:
```javascript
const depositBox = new Proxy({}, {
	get() { throw new Error('Deposits only, please') }
})
```

[Promises/A+ understands the risk of such objects](https://promisesaplus.com/#the-promise-resolution-procedure) as well:
 > If retrieving the property `x.then` results in a thrown exception `e`, reject `promise` with `e` as the reason.

### Issue

We're using _ruply_ in projects in which performance matters, and it seems the try-catch is causing a significant slowdown.

We should figure out:
1. What are the requirements for a `checkThenable` function?
2. How fast or slow are the different alternatives compared to the current implementation?

_is-promise_, a library designed to do the same thing as `checkThenable`, [had a discussion concerning "dangerous objects"](https://github.com/then/is-promise/pull/4) seven years ago. They decided against catching the error. I'm not completely sure whether it's a fair comparison, though. Having `isPromise(dangerousObject)` cause an error is maybe less surprising than having `run(dangerousObject, […])` cause an error.

### PR DESCRIPTION
**Pull request by janpaepke**
This resolves https://github.com/Pimm/ruply/issues/2

### CODE REVIEW

File: test/basics.test.js [hunk 1]
```diff
@@ -70,4 +70,23 @@ test('basics', () => {
...
+		const throwingGetter = {
+			get then() { throw new Error() }
+		}
+		expect(
+			() => apply(throwingGetter, callback)
+		).not.toThrow();
+		expect(callback).toHaveBeenCalledWith(throwingGetter);
+	});
+	// apply with object, where any getter throws.
+	run(createCallback(), callback => {
```
Comment (janpaepke) [inline test/basics.test.js]: I couldn't test those in one go, as it turns out using `expect(callback)` with the proxy that always fails, calls the getters and thus breaks. Likely this is a jest feature getting info on the passed args.

Comment (Pimm) [reply test/basics.test.js]: `expect` blows up with the hostile proxy as well, I see!

I worked around this by making _just_ `then` throw.

**Comment by Pimm** [comment]
Semi-interesting fact: this PR does not change the size of the compiled version one byte (before GZIP, that is).

**Comment by janpaepke** [comment]
I'm down with all the changes.

Was considering writing this instead, as I find it cleaner than reassigning parameters. It's also more consistent with the other use in `build()`.
```js
	apply =
	build(
		'apply',
		function applyLogic(value, callback) {
			// Call the callback and check whether the result is promise-like.
			const then = takeThen(callback.call(this, value))
			return then
				// If the result is promise-like chain a function to it which will return the value, then return that chain.
				? then(() => value)
				// If the result is not promise-like, return the value.
				: value;
		}
	);
```

But as it turns out, this adds 2 bytes and I guess it's not worth breaking your semi-interesting fact.

**Comment by janpaepke** [comment]
Good to merge?

**Pull request merged**